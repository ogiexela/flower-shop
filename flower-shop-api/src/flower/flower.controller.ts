import { Controller } from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { Crud } from '@nestjsx/crud';
import { FlowerEntity } from './flower.entity';
import { FlowerService } from './flower.service';

@Crud({
  model: {
    type: FlowerEntity
  },
  params: {
    id: {
      field: 'id',
      type: 'uuid',
      primary: true
    }
  }
})
@Controller('flower')
@ApiTags('flower')
export class FlowerController {
  constructor(public service: FlowerService) { }
}