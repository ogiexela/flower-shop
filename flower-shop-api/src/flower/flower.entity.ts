import {
  Column,
  Entity,
  PrimaryColumn,
  PrimaryGeneratedColumn,
} from 'typeorm';

@Entity('flower')
export class FlowerEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @PrimaryColumn()
  sku: string;

  @Column({ length: 100 })
  name: string;

  @Column({ length: 100 })
  image: string;

  @Column('float')
  price: number;

  @Column('int', { default: 0 })
  rating: number;
}