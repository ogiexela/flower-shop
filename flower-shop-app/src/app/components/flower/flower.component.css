
.flower-card {
    border-radius: 5px 5px 5px 5px;
    box-shadow: 0 0 5px rgb(189, 190, 190);
    display: flex;
    flex-direction: column;
    margin: 20px;
    max-width: 200px;
    padding: 10px;
    position: relative;
    text-align: center;
}

button {
    background-color: #42e695;
    border: none;
    color: white;
    cursor: pointer;
    height: 25px;
    left: 50%;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    width: 150px;
}

.flower-card img {
    border-bottom: solid 1px lightgray;
    margin-bottom: 10px;
    margin: 5x;
    padding-bottom: 10px;
    width: 190px;
    height: 180px;
}

.flower-price {
    color: grey;
}

.flower-rating span {
    margin: 2px;
}

.half-opacity {
    opacity: 0.5;
}

.flower-image {
    position: relative;
}

.in-cart {
    background-color: #ff6347;
    border-radius: 50%;
    color: white;
    font-size: small;
    height: 50px;
    position: absolute;
    text-align: center;
    vertical-align: middle;
    width: 50px;
    top: -8px;
    left: -8px;
}

.flower-name {
    font-weight: bold;
}

.in-cart span {
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    width: 100%;
}