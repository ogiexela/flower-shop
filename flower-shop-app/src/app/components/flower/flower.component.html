<div
    (mouseenter)="setAddToCart(true)"
    (mouseleave)="setAddToCart(false)"
    *ngIf="flower"
    class="flower-card"
    id="flower_card_{{flower?.id}}"
>
    <div class="flower-image">
        <img
            [ngClass]="{'half-opacity': inCart || showAddToCart}"
            src="./assets/flowers/{{flower.image}}"
        >
        <button
            (click)="onToCart()"
            *ngIf="showAddToCart && !inCart"
            id="add_flower_to_cart"
        >Add to cart</button>
        <button
            (click)="onRemoveFromCart()"
            *ngIf="inCart"
            id="remove_flower_from_cart"
        >Remove from cart</button>
    </div>
    <span class="flower-name">{{flower.name}}</span>
    <span class="flower-price">${{price}}</span>
    <div class="flower-rating">
        <span
            *ngFor="let star of rating"
            [innerHTML]="star"
        ></span>
    </div>
    <div
        *ngIf="inCart"
        class="in-cart"
    >
        <span>In Cart</span>
    </div>
</div>
<div *ngIf="!flower">404 Flower not found.</div>
