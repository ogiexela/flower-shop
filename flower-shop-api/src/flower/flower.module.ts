import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { FlowerController } from './flower.controller';
import { FlowerEntity } from './flower.entity';
import { FlowerService } from './flower.service';

@Module({
  imports: [TypeOrmModule.forFeature([FlowerEntity])],
  controllers: [FlowerController],
  providers: [FlowerService]
})
export class FlowerModule { }