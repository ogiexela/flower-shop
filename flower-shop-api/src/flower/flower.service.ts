import { Repository } from 'typeorm';
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { FlowerEntity } from './flower.entity';

@Injectable()
export class FlowerService extends TypeOrmCrudService<FlowerEntity> {
  constructor(@InjectRepository(FlowerEntity) repo: Repository<FlowerEntity>) {
    super(repo)
  }
}