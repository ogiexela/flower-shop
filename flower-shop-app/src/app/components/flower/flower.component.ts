import {
  Component,
  EventEmitter,
  Input,
  OnChanges,
  OnInit,
  Output,
} from '@angular/core';
import { Flower } from '../../models/flower.model';

@Component({
  selector: 'app-flower',
  templateUrl: './flower.component.html',
  styleUrls: ['./flower.component.css']
})
export class FlowerComponent implements OnInit, OnChanges {

  @Input() public flower: Flower;
  @Input() public inCart = false;
  @Output() public addToCart: EventEmitter<string> = new EventEmitter<string>();
  @Output() public removeFromCart: EventEmitter<string> = new EventEmitter<string>();

  public rating: string[] = Array(5).fill('&#9734;');
  public price: string;
  public showCartControls = false;

  ngOnInit() {
    this.updateInfo();
  }

  ngOnChanges(): void {
   this.updateInfo();
  }

  private updateInfo(): void {
    if (!this.flower) {
      return;
    }


    this.rating = [...this.rating].fill('&#9733', 0, this.flower.rating);
    this.price = `${this.flower.price.toFixed(2)}`;
  }

  public setShowCartControls(value: boolean): void {
    this.showCartControls = value;
  }

  public onAddToCart(): void {
    this.inCart = true;
    this.addToCart.emit(this.flower.id);
  }

  public onRemoveFromCart(): void {
    this.inCart = false;
    this.removeFromCart.emit(this.flower.id);
  }

  public get showAddToCartButton(): boolean {
    return this.showCartControls && !this.inCart;
  }

  public get showRemoveFromCartButton(): boolean {
    return this.showCartControls && this.inCart;
  }
}
