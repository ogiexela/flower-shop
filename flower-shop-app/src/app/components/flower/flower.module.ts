import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FlowerComponent } from './flower.component';

@NgModule({
  declarations: [FlowerComponent],
  imports: [
    CommonModule
  ],
  exports: [
    FlowerComponent,
  ]
})
export class FlowerModule { }
