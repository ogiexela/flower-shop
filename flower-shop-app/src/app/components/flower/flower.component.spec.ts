import { Component } from '@angular/core';
import {
  async,
  ComponentFixture,
  TestBed,
} from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { Flower } from '../../models/flower.model';
import { FlowerModule } from './flower.module';

@Component({
  template: `
    <app-flower
      (addToCart)="addToCart($event)"
      (removeFromCart)="removeFromCart($event)"
      [flower]="flower"
      [inCart]="inCart"
    ></app-flower>`
})
class TestHostComponent {
  public flower: Flower;
  public inCart = false;
  public addToCart(): void {}
  public removeFromCart(): void {}
}

describe('FlowerComponent', () => {
  let fixture: ComponentFixture<TestHostComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        FlowerModule,
      ],
      declarations: [TestHostComponent],
      providers: [],
    }).compileComponents()
      .then(() => {
        fixture = TestBed.createComponent(TestHostComponent);
        fixture.detectChanges();
      })
      .catch(error => console.error(error));
  }));

  it('state without a flower', () => {
    expect(fixture.nativeElement).toMatchSnapshot('empty_state');
  });

  it('state with a flower', async () => {
    fixture.componentInstance.flower = {
      name: 'Test Flower1',
      id: 'fooid',
      image: 'flower1.jpg',
      price: 4.99,
      rating: 2,
      sku: 'tf1',
    };

    fixture.detectChanges();

    await fixture.whenRenderingDone();

    expect(fixture.nativeElement).toMatchSnapshot('flower_state');
  });

  it('display add to cart on mouse enter', async () => {
    fixture.componentInstance.flower = {
      id: 'fooid',
      image: 'flower2.jpg',
      name: 'Test Flower2',
      price: 4.99,
      rating: 3,
      sku: 'tf1',
    };

    fixture.detectChanges();

    const flowerCard = fixture.debugElement.query(By.css('.flower-card'));
    const mouseenter = new MouseEvent('mouseenter');
    flowerCard.nativeElement.dispatchEvent(mouseenter);

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot('flower_add_to_cart');
  });

  it('do not display add to cart on mouse leave', async () => {
    fixture.componentInstance.flower = {
      id: 'fooid',
      image: 'flower3.jpg',
      name: 'Test Flower3',
      price: 5.99,
      rating: 1,
      sku: 'tf3',
    };

    fixture.detectChanges();

    const flowerCard = fixture.debugElement.query(By.css('.flower-card'));
    const mouseenter = new MouseEvent('mouseenter');
    flowerCard.nativeElement.dispatchEvent(mouseenter);

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot('flower_add_to_cart_on');

    const mouseleave = new MouseEvent('mouseleave');
    flowerCard.nativeElement.dispatchEvent(mouseleave);

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot('flower_add_to_cart_off');
  });

  it('able to add to cart', async () => {
    const testFlower: Flower = {
      id: 'foo_id1',
      name: 'Test Flower4',
      image: 'flower4.jpg',
      sku: 'tf4',
      price: 5.99,
      rating: 4,
    };

    fixture.componentInstance.flower = testFlower;

    fixture.detectChanges();

    const flowerCard = fixture.debugElement.query(By.css('.flower-card'));
    const mouseenter = new MouseEvent('mouseenter');
    flowerCard.nativeElement.dispatchEvent(mouseenter);

    fixture.detectChanges();

    const addToCartSpy = jest.spyOn(fixture.componentInstance, 'addToCart');

    const addToCart = fixture.debugElement.query(By.css('#add_flower_to_cart'));
    const click = new MouseEvent('click');
    addToCart.nativeElement.dispatchEvent(click);

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot('flower_added_to_cart');

    expect(addToCartSpy).toBeCalledWith(testFlower.id);

    addToCartSpy.mockClear();
  });

  it('able to remove from cart', async () => {
    const testFlower: Flower = {
      id: 'foo_id2',
      name: 'Test Flower5',
      image: 'flower5.jpg',
      sku: 'tf5',
      price: 6.99,
      rating: 5,
    };
    fixture.componentInstance.inCart = true;
    fixture.componentInstance.flower = testFlower;

    fixture.detectChanges();

    const removeFromCartSpy = jest.spyOn(fixture.componentInstance, 'removeFromCart');

    const removeFromCart = fixture.debugElement.query(By.css('#remove_flower_from_cart'));
    const click = new MouseEvent('click');

    removeFromCart.nativeElement.dispatchEvent(click);

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot('flower_added_to_cart_false');

    expect(removeFromCartSpy).toBeCalledWith(testFlower.id);

    removeFromCartSpy.mockClear();
  });

  it('flower already in cart', async () => {
    fixture.componentInstance.inCart = true;
    fixture.componentInstance.flower = {
      id: 'fooid',
      image: 'flower7.jpg',
      name: 'Test Flower7',
      price: 7.99,
      rating: 3,
      sku: 'tf7',
    };

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot('flower_already_in_cart');

    const flowerCard = fixture.debugElement.query(By.css('.flower-card'));
    const mouseenter = new MouseEvent('mouseenter');
    flowerCard.nativeElement.dispatchEvent(mouseenter);

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot('flower_already_in_cart');
  });
});
