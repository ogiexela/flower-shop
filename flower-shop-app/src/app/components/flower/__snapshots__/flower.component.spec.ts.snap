// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowerComponent able to add to cart: flower_added_to_cart 1`] = `
<div
  id="root4"
>
  <app-flower>
    
    <div
      class="flower-card"
      id="flower_card_foo_id1"
    >
      <div
        class="flower-image"
      >
        <img
          class="half-opacity"
          src="./assets/flowers/flower4.jpg"
        />
        
        
        <button
          id="remove_flower_from_cart"
        >
          Remove from cart
        </button>
      </div>
      <span
        class="flower-name"
      >
        Test Flower4
      </span>
      <span
        class="flower-price"
      >
        $5.99
      </span>
      <div
        class="flower-rating"
      >
        
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ☆
        </span>
      </div>
      
      <div
        class="in-cart"
      >
        <span>
          In Cart
        </span>
      </div>
    </div>
    
  </app-flower>
</div>
`;

exports[`FlowerComponent able to remove from cart: flower_added_to_cart_false 1`] = `
<div
  id="root5"
>
  <app-flower>
    
    <div
      class="flower-card"
      id="flower_card_foo_id2"
    >
      <div
        class="flower-image"
      >
        <img
          class=""
          src="./assets/flowers/flower5.jpg"
        />
        
        
      </div>
      <span
        class="flower-name"
      >
        Test Flower5
      </span>
      <span
        class="flower-price"
      >
        $6.99
      </span>
      <div
        class="flower-rating"
      >
        
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ★
        </span>
      </div>
      
    </div>
    
  </app-flower>
</div>
`;

exports[`FlowerComponent display add to cart on mouse enter: flower_add_to_cart 1`] = `
<div
  id="root2"
>
  <app-flower>
    
    <div
      class="flower-card"
      id="flower_card_fooid"
    >
      <div
        class="flower-image"
      >
        <img
          class="half-opacity"
          src="./assets/flowers/flower2.jpg"
        />
        
        <button
          id="add_flower_to_cart"
        >
          Add to cart
        </button>
        
      </div>
      <span
        class="flower-name"
      >
        Test Flower2
      </span>
      <span
        class="flower-price"
      >
        $4.99
      </span>
      <div
        class="flower-rating"
      >
        
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
      </div>
      
    </div>
    
  </app-flower>
</div>
`;

exports[`FlowerComponent do not display add to cart on mouse leave: flower_add_to_cart_off 1`] = `
<div
  id="root3"
>
  <app-flower>
    
    <div
      class="flower-card"
      id="flower_card_fooid"
    >
      <div
        class="flower-image"
      >
        <img
          class=""
          src="./assets/flowers/flower3.jpg"
        />
        
        
      </div>
      <span
        class="flower-name"
      >
        Test Flower3
      </span>
      <span
        class="flower-price"
      >
        $5.99
      </span>
      <div
        class="flower-rating"
      >
        
        <span>
          ★
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
      </div>
      
    </div>
    
  </app-flower>
</div>
`;

exports[`FlowerComponent do not display add to cart on mouse leave: flower_add_to_cart_on 1`] = `
<div
  id="root3"
>
  <app-flower>
    
    <div
      class="flower-card"
      id="flower_card_fooid"
    >
      <div
        class="flower-image"
      >
        <img
          class="half-opacity"
          src="./assets/flowers/flower3.jpg"
        />
        
        <button
          id="add_flower_to_cart"
        >
          Add to cart
        </button>
        
      </div>
      <span
        class="flower-name"
      >
        Test Flower3
      </span>
      <span
        class="flower-price"
      >
        $5.99
      </span>
      <div
        class="flower-rating"
      >
        
        <span>
          ★
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
      </div>
      
    </div>
    
  </app-flower>
</div>
`;

exports[`FlowerComponent flower already in cart: flower_already_in_cart 1`] = `
<div
  id="root6"
>
  <app-flower>
    
    <div
      class="flower-card"
      id="flower_card_fooid"
    >
      <div
        class="flower-image"
      >
        <img
          class="half-opacity"
          src="./assets/flowers/flower7.jpg"
        />
        
        
        <button
          id="remove_flower_from_cart"
        >
          Remove from cart
        </button>
      </div>
      <span
        class="flower-name"
      >
        Test Flower7
      </span>
      <span
        class="flower-price"
      >
        $7.99
      </span>
      <div
        class="flower-rating"
      >
        
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
      </div>
      
      <div
        class="in-cart"
      >
        <span>
          In Cart
        </span>
      </div>
    </div>
    
  </app-flower>
</div>
`;

exports[`FlowerComponent flower already in cart: flower_already_in_cart 2`] = `
<div
  id="root6"
>
  <app-flower>
    
    <div
      class="flower-card"
      id="flower_card_fooid"
    >
      <div
        class="flower-image"
      >
        <img
          class="half-opacity"
          src="./assets/flowers/flower7.jpg"
        />
        
        
        <button
          id="remove_flower_from_cart"
        >
          Remove from cart
        </button>
      </div>
      <span
        class="flower-name"
      >
        Test Flower7
      </span>
      <span
        class="flower-price"
      >
        $7.99
      </span>
      <div
        class="flower-rating"
      >
        
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
      </div>
      
      <div
        class="in-cart"
      >
        <span>
          In Cart
        </span>
      </div>
    </div>
    
  </app-flower>
</div>
`;

exports[`FlowerComponent state with a flower: flower_state 1`] = `
<div
  id="root1"
>
  <app-flower>
    
    <div
      class="flower-card"
      id="flower_card_fooid"
    >
      <div
        class="flower-image"
      >
        <img
          src="./assets/flowers/flower1.jpg"
        />
        
        
      </div>
      <span
        class="flower-name"
      >
        Test Flower1
      </span>
      <span
        class="flower-price"
      >
        $4.99
      </span>
      <div
        class="flower-rating"
      >
        
        <span>
          ★
        </span>
        <span>
          ★
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
        <span>
          ☆
        </span>
      </div>
      
    </div>
    
  </app-flower>
</div>
`;

exports[`FlowerComponent state without a flower: empty_state 1`] = `
<div
  id="root0"
>
  <app-flower>
    
    
    <div>
      404 Flower not found.
    </div>
  </app-flower>
</div>
`;
